From e13b55915080911719c4f917bccc44fb7016cc14 Mon Sep 17 00:00:00 2001
From: Alisa Hung <alisahung@google.com>
Date: Fri, 13 Jun 2025 05:52:18 -0700
Subject: [PATCH] Trim oversized strings in setId and setConversationId

Flag: EXEMPT bugfix
Test: android.app.NotificationChannelTest
Test: cts NotificationChannelTest
Bug: 419014146
Change-Id: Id721d3123ee8d38753f550fe57ba0f5d15d743ac
(cherry picked from commit a9a1b4365b8d6def5b36fed7dd9caf2389fe0a51)
---
 .../java/android/app/NotificationChannel.java |  6 ++--
 .../android/app/NotificationChannelTest.java  | 30 +++++++++++++++++++
 2 files changed, 33 insertions(+), 3 deletions(-)

diff --git a/core/java/android/app/NotificationChannel.java b/core/java/android/app/NotificationChannel.java
index 73d26b834497..2ca6e97a8399 100644
--- a/core/java/android/app/NotificationChannel.java
+++ b/core/java/android/app/NotificationChannel.java
@@ -621,7 +621,7 @@ public final class NotificationChannel implements Parcelable {
      * @hide
      */
     public void setId(String id) {
-        mId = id;
+        mId = getTrimmedString(id);
     }
 
     // Modifiable by apps on channel creation.
@@ -865,8 +865,8 @@ public final class NotificationChannel implements Parcelable {
      */
     public void setConversationId(@NonNull String parentChannelId,
             @NonNull String conversationId) {
-        mParentId = parentChannelId;
-        mConversationId = conversationId;
+        mParentId = getTrimmedString(parentChannelId);
+        mConversationId = getTrimmedString(conversationId);
     }
 
     /**
diff --git a/core/tests/coretests/src/android/app/NotificationChannelTest.java b/core/tests/coretests/src/android/app/NotificationChannelTest.java
index e4b54071e892..cfe9b7afef54 100644
--- a/core/tests/coretests/src/android/app/NotificationChannelTest.java
+++ b/core/tests/coretests/src/android/app/NotificationChannelTest.java
@@ -235,6 +235,36 @@ public class NotificationChannelTest {
                 fromParcel.getConversationId().length());
     }
 
+    @Test
+    public void testSetId_longStringIsTrimmed() {
+        NotificationChannel channel =
+                new NotificationChannel("id", "name", NotificationManager.IMPORTANCE_DEFAULT);
+        String longId = Strings.repeat("A", NotificationChannel.MAX_TEXT_LENGTH + 10);
+
+        channel.setId(longId);
+
+        assertThat(channel.getId()).hasLength(NotificationChannel.MAX_TEXT_LENGTH);
+        assertThat(channel.getId())
+                .isEqualTo(longId.substring(0, NotificationChannel.MAX_TEXT_LENGTH));
+    }
+
+    @Test
+    public void testSetConversationId_longStringsAreTrimmed() {
+        NotificationChannel channel =
+                new NotificationChannel("id", "name", NotificationManager.IMPORTANCE_DEFAULT);
+        String longParentId = Strings.repeat("P", NotificationChannel.MAX_TEXT_LENGTH + 10);
+        String longConversationId = Strings.repeat("C", NotificationChannel.MAX_TEXT_LENGTH + 10);
+
+        channel.setConversationId(longParentId, longConversationId);
+
+        assertThat(channel.getParentChannelId()).hasLength(NotificationChannel.MAX_TEXT_LENGTH);
+        assertThat(channel.getParentChannelId())
+                .isEqualTo(longParentId.substring(0, NotificationChannel.MAX_TEXT_LENGTH));
+        assertThat(channel.getConversationId()).hasLength(NotificationChannel.MAX_TEXT_LENGTH);
+        assertThat(channel.getConversationId())
+                .isEqualTo(longConversationId.substring(0, NotificationChannel.MAX_TEXT_LENGTH));
+    }
+
     @Test
     public void testLongAlertFields() {
         NotificationChannel channel = new NotificationChannel("id", "name", 3);
-- 
2.34.1

